packages/firebase_messaging/firebase_messaging/android/src/main/java/io/flutter/plugins/firebase/messaging/JobIntentService.java

```diff
@@ -342,32 +344,42 @@ public void complete() {
   }
 
   /** This is a task to dequeue and process work in the background. */
-  final class CommandProcessor extends AsyncTask<Void, Void, Void> {
-    @Override
-    protected Void doInBackground(Void... params) {
-      GenericWorkItem work;
-
-      if (DEBUG) Log.d(TAG, "Starting to dequeue work...");
-
-      while ((work = dequeueWork()) != null) {
-        if (DEBUG) Log.d(TAG, "Processing next work: " + work);
-        onHandleWork(work.getIntent());
-        if (DEBUG) Log.d(TAG, "Completing work: " + work);
-        work.complete();
-      }
-
-      if (DEBUG) Log.d(TAG, "Done processing work!");
-
-      return null;
-    }
-
-    @Override
-    protected void onCancelled(Void aVoid) {
-      processorFinished();
+  final class CommandProcessor {
+    private final Executor executor = Executors.newSingleThreadExecutor(); // Background thread
+    private final Handler handler = new Handler(Looper.getMainLooper()); // UI Thread
+
+    public void execute() {
+      executor.execute(
+          new Runnable() {
```

Comment:
Should we store this runnable to cancel it manually inside `cancel()` if needed?

Reply:
I don't think it is necessary, it is a like for like replacement replacing deprecated AsyncTask. The execute(Runnable) is running in the background thread (like `doInBackground` for AyncTask). When that work is complete, it calls the `processorFinished()` finished on the main thread (like `onPostExecute` for AsyncTask). See documentation: https://developer.android.com/reference/android/os/AsyncTask#the-4-steps

Reply:
To add to that, the only time cancel is called using previous AsyncTask, we also call cancel() here using latest implementation: https://github.com/firebase/flutterfire/pull/12580/files#diff-3d9ef1ece89413638d5bc6907c220242d3b2f489db5900782615c504073437f1R561